docs/explore/calibanconfig.rst
calibanconfig
^^^^^^^^^^^^^^^^^^^^^^

Caliban supports customization through a file called ``.calibanconfig.json``
that lives in your project's directory. Features are limited for now, but stay
tuned for more.

Custom Apt Packages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Caliban provides support for custom aptitude packages inside your container. To
require custom apt packages, create a file called ``.calibanconfig.json`` inside
your project's directory.

The ``.calibanconfig.json`` should contain a single JSON dictionary with an
``"apt_packages"`` key. The value under this key can be either a list, or a
dictionary with ``"gpu"`` and ``"cpu"'`` keys. For example, any of the following are
valid:

.. code-block::

   # This is a list by itself. Comments are fine, by the way.
   {
        "apt_packages": ["libsm6", "libxext6", "libxrender-dev"]
   }

This works too:

.. code-block:: json

   # You can also include a dictionary with different deps
   # for gpu and cpu modes. It's fine to leave either of these blank,
   # or not include it.
   {
       "apt_packages": {
           "gpu": ["libsm6", "libxext6", "libxrender-dev"],
           "cpu": ["some_other_package"]
       }
   }

These values will do what you expect and run ``apt-get install <package_name>``
for each package. Packages are alphabetized, so changing the order won't
invalidate Docker's build cache.

Custom Base Images
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
For details on Caliban's base images, see :ref:`What's the Base Docker Image?`.

You can specify a custom base image for Caliban to use in your ``.calibanconfig.json`` file
by adding an entry with the ``base_image`` key as follows:

.. code-block:: json

   {
       "base_image": "gcr.io/blueshift-playground/blueshift:gpu-ubuntu1804-py38-cuda101"
   }

You can also specify different base images for ``cpu`` and ``gpu`` modes as follows:

.. code-block:: json

   {
       "base_image": {
           "cpu": "gcr.io/blueshift-playground/blueshift:cpu-ubuntu1804-py38",
           "gpu": "gcr.io/blueshift-playground/blueshift:gpu-ubuntu1804-py38-cuda101"
       }
   }


CHANGELOG.md
# pending

# 0.4.1

Small release to archive for JOSS acceptance.

- Move `cloud_sql_proxy` installation before code copy (https://github.com/google/caliban/pull/87)

# 0.4.0

The biggest feature in this new release is native support for logging to an
MLFlow tracking server using the [UV
Metrics](http://github.com/google/uv-metrics) project.
(https://github.com/google/caliban/pull/35) This feature is in alpha; expect
documentation soon.

### More features

- minor bugfixes for GKE (https://github.com/google/caliban/pull/85)
- additional tests for gke.{types, util} (https://github.com/google/caliban/pull/84)
- re-order custom apt packages before pip requirements (https://github.com/google/caliban/pull/82)
- modify base image to our more general cloudbuild naming scheme (https://github.com/google/caliban/pull/80)
- updated `google-auth` dependency version to `1.19.0` (https://github.com/google/caliban/pull/79)
- add clearer contribution info (https://github.com/google/caliban/pull/76)
- Update uv-metrics tutorial (https://github.com/google/caliban/pull/74, https://github.com/google/caliban/pull/72)
- add support for running an embedded cloudsql_proxy (https://github.com/google/caliban/pull/60)
- bugfix for #65: do not add resource maxima when quota is < 1 (#67)
- Updated accelerator regions (and globally availabe AI Platform regions to
  match the current state here):
  https://cloud.google.com/ai-platform/training/docs/regions

# 0.3.0

- @ramasesh Added a fix that prevented `pip` git dependencies from working in
  `caliban shell` mode (https://github.com/google/caliban/pull/55) This adds a
  small update to the base image, so be sure to run

```
docker pull gcr.io/blueshift-playground/blueshift:cpu
docker pull gcr.io/blueshift-playground/blueshift:gpu
```

to get access to this fix.

- Thanks to @eschnett, `--docker_run-args` can now deal with arbitrary
  whitespace in the list of arguments, instead of single spaces only.
  (https://github.com/google/caliban/pull/46)

- Caliban now authenticates AI Platform job submissions using the authentication
  provided by `gcloud auth login`, rather than requiring a service account key.
  This significantly simplifies the setup required for a first time user.

- `caliban cloud` now checks if the image exists remotely before issuing a
  `docker push` command on the newly built image
  (https://github.com/google/caliban/pull/36)

- Big internal refactor to make it easier to work on code, increase test
  coverage, add new backends (https://github.com/google/caliban/pull/32)

- add `schema` validation for `.calibanconfig.json`. This makes it much easier
  to add configuration knobs: https://github.com/google/caliban/pull/37

- Custom base image support (https://github.com/google/caliban/pull/39), thanks
  to https://github.com/google/caliban/pull/20 from @sagravat.
  `.calibanconfig.json` now supports a `"base_image"` key. For the value, can
  supply:
  - a Docker base image of your own
  - a dict of the form `{"cpu": "base_image", "gpu": "base_image"}` with both
    entries optional, of course.

  Two more cool features.

  First, if you use a format string, like `"my_image-{}:latest"`, the format
  block `{}` will be filled in with either `cpu` or `gpu`, depending on the mode
  Caliban is using.

  Second, we now have native support for [Google's Deep Learning
  VMs](https://cloud.google.com/ai-platform/deep-learning-vm/docs/introduction)
  as base images. The actual VM containers [live
  here](https://console.cloud.google.com/gcr/images/deeplearning-platform-release/GLOBAL).
  If you provide any of the following strings, Caliban will expand them out to
  the actual base image location:

```
dlvm:pytorch-cpu
dlvm:pytorch-cpu-1.0
dlvm:pytorch-cpu-1.1
dlvm:pytorch-cpu-1.2
dlvm:pytorch-cpu-1.3
dlvm:pytorch-cpu-1.4
dlvm:pytorch-gpu
dlvm:pytorch-gpu-1.0
dlvm:pytorch-gpu-1.1
dlvm:pytorch-gpu-1.2
dlvm:pytorch-gpu-1.3
dlvm:pytorch-gpu-1.4
dlvm:tf-cpu
dlvm:tf-cpu-1.0
dlvm:tf-cpu-1.13
dlvm:tf-cpu-1.14
dlvm:tf-cpu-1.15
dlvm:tf-gpu
dlvm:tf-gpu-1.0
dlvm:tf-gpu-1.13
dlvm:tf-gpu-1.14
dlvm:tf-gpu-1.15
dlvm:tf2-cpu
dlvm:tf2-cpu-2.0
dlvm:tf2-cpu-2.1
dlvm:tf2-cpu-2.2
dlvm:tf2-gpu
dlvm:tf2-gpu-2.0
dlvm:tf2-gpu-2.1
dlvm:tf2-gpu-2.2
```

Format strings work here as well! So, `"dlvm:pytorch-{}-1.4"` is a totally valid
base image.

# 0.2.6

- Prepared for a variety of base images by setting up a cloud build matrix:
  https://github.com/google/caliban/pull/25
- Added better documentation for `gcloud auth configure-docker`
  https://github.com/google/caliban/pull/26
- Added `close()` to `TqdmFile`, preventing an error when piping `stdout`:
  https://github.com/google/caliban/pull/30
- `tqdm` progress bars and other interactive outputs now display correctly in
  `caliban run` outputs. `stdout` flushes properly! Before these changes,
  `stderr` would appear before any `stdout`, making it difficult to store the
  logs in a text file. Now, by default, python processes launched by `caliban
  run` won't buffer. https://github.com/google/caliban/pull/31

![2020-06-26 09 48 50](https://user-images.githubusercontent.com/69635/85877300-2a3e7300-b794-11ea-9792-4cf3ae5e4263.gif)

# 0.2.5

- fixes the python binary that caliban notebook points to (now that we use
  conda)
- adds DEBIAN_FRONTEND=noninteractive to the apt-get command, so that packages
  like texlive won't freeze and wait for you to specify a timezone.

This makes it easy to add, for example, npm and latex support to your caliban
notebook invocations.

# 0.2.4

- fixes a bug with `parse_region` not handling a lack of default.
- converts the build to Github Actions.
- Rolls Caliban back to requiring only python 3.6 support.
- Removes some unused imports from a few files.

# 0.2.3

- Added fix for an issue where large user IDs would crash Docker during the
  build phase. https://github.com/google/caliban/pull/8

# 0.2.2

- Fix for bug with requirements.txt files.

# 0.2.1

- Added support for Conda dependencies

caliban/config.py
#!/usr/bin/python
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Utilities for our job runner, for working with configs.
"""

from __future__ import absolute_import, division, print_function

import argparse
import itertools
import os
import sys
from enum import Enum
from typing import Any, Dict, List, Union, Optional
import re
import commentjson
import yaml

import caliban.cloud.types as ct
import caliban.util as u

# int, str and bool are allowed in a final experiment; lists are markers for
# expansion.
ExpValue = Union[int, str, bool]

# Entry in an experiment config. If any values are lists they're expanded into
# a sequence of experiment configs.
Expansion = Dict[str, Union[ExpValue, List[ExpValue]]]

# An experiment config can be a single (potentially expandable) dictionary, or
# a list of many such dicts.
ExpConf = Union[Expansion, List[Expansion]]

# A final experiment can only contain valid ExpValues, no expandable entries.
Experiment = Dict[str, ExpValue]


# Mode
class JobMode(str, Enum):
  CPU = 'CPU'
  GPU = 'GPU'


# Special config for Caliban.
CalibanConfig = Dict[str, Any]

DRY_RUN_FLAG = "--dry_run"
CALIBAN_CONFIG = ".calibanconfig.json"

# Defaults for various input values that we can supply given some partial set
# of info from the CLI.
DEFAULT_REGION = ct.US.central1

# : Dict[JobMode, ct.MachineType]
DEFAULT_MACHINE_TYPE = {
    JobMode.CPU: ct.MachineType.highcpu_32,
    JobMode.GPU: ct.MachineType.standard_8
}
DEFAULT_GPU = ct.GPU.P100

# Config to supply for CPU jobs.
DEFAULT_ACCELERATOR_CONFIG = {
    "count": 0,
    "type": "ACCELERATOR_TYPE_UNSPECIFIED"
}


def extract_dlvm_image(job_mode: JobMode, dlvm_arg: str) -> str:
  """Returns the DLVM image url for the job model and the comand line arg
  or returns None if the key doesn't exist in the config.

  """
  return _dlvm_config(job_mode).get(dlvm_arg)


def _dlvm_config(job_mode: JobMode) -> Dict[str, str]:
  job_mode_str = job_mode.lower()
  return {
    "pytorch": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}",
    "pytorch-1.0": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}.1-0",
    "pytorch-1.1": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}.1-1",
    "pytorch-1.2": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}.1-2",
    "pytorch-1.3": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}.1-3",
    "pytorch-1.4": f"gcr.io/deeplearning-platform-release/pytorch-{job_mode_str}.1-4",
    "tf": f"gcr.io/deeplearning-platform-release/tf-{job_mode_str}",
    "tf-1.13": f"gcr.io/deeplearning-platform-release/tf-{job_mode_str}.1-13",
    "tf-1.14": f"gcr.io/deeplearning-platform-release/tf-{job_mode_str}.1-14",
    "tf-1.15": f"gcr.io/deeplearning-platform-release/tf-{job_mode_str}.1-15",
    "tf2": f"gcr.io/deeplearning-platform-release/tf2-{job_mode_str}",
    "tf-2.0": f"gcr.io/deeplearning-platform-release/tf2-{job_mode_str}.2-0",
    "tf-2.1": f"gcr.io/deeplearning-platform-release/tf2-{job_mode_str}.2-1",
    "tf-2.2": f"gcr.io/deeplearning-platform-release/tf2-{job_mode_str}.2-2"
  }

def gpu(job_mode: JobMode) -> bool:
  """Returns True if the supplied JobMode is JobMode.GPU, False otherwise.

  """
  return job_mode == JobMode.GPU


def load_yaml_config(path):
  """returns the config parsed based on the info in the flags.

  Grabs the config file, written in yaml, slurps it in.
  """
  with open(path) as f:
    config = yaml.load(f, Loader=yaml.FullLoader)

  return config


def load_config(path, mode='yaml'):
  """Load a JSON or YAML config.

  """
  if mode == 'json':
    with open(path) as f:
      return commentjson.load(f)

  return load_yaml_config(path)


def valid_json(path: str) -> Dict[str, Any]:
  """Loads JSON if the path points to a valid JSON file; otherwise, throws an
  exception that's picked up by argparse.

  """
  try:
    return load_config(path, mode='json')
  except commentjson.JSONLibraryException:
    raise argparse.ArgumentTypeError(
        """File '{}' doesn't seem to contain valid JSON. Try again!""".format(
            path))


def extract_script_args(m: Dict[str, Any]) -> List[str]:
  """Strip off the "--" argument if it was passed in as a separator."""
  script_args = m.get("script_args")
  if script_args is None or script_args == []:
    return script_args

  head, *tail = script_args

  return tail if head == "--" else script_args


def extract_project_id(m: Dict[str, Any]) -> str:
  """Attempts to extract the project_id from the args; falls back to an
  environment variable, or exits if this isn't available. There's no sensible
  default available.

  """
  project_id = m.get("project_id") or os.environ.get("PROJECT_ID")

  if project_id is None:
    print()
    print(
        "\nNo project_id found. 'caliban cloud' requires that you either set a \n\
$PROJECT_ID environment variable with the ID of your Cloud project, or pass one \n\
explicitly via --project_id. Try again, please!")
    print()

    sys.exit(1)

  return project_id


def extract_region(m: Dict[str, Any]) -> ct.Region:
  """Returns the region specified in the args; defaults to an environment
  variable. If that's not supplied defaults to the default cloud provider from
  caliban.cloud.

  """
  region = m.get("region") or os.environ.get("REGION")

  if region:
    return ct.parse_region(region)

  return DEFAULT_REGION


def extract_zone(m: Dict[str, Any]) -> str:
  return "{}-a".format(extract_region(m))


def extract_cloud_key(m: Dict[str, Any]) -> Optional[str]:
  """Returns the Google service account key filepath specified in the args;
  defaults to the $GOOGLE_APPLICATION_CREDENTIALS variable.

  """
  return m.get("cloud_key") or \
    os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")


def apt_packages(conf: CalibanConfig, mode: JobMode) -> List[str]:
  """Returns the list of aptitude packages that should be installed to satisfy
  the requests in the config.

  """
  packages = conf.get("apt_packages") or {}

  if isinstance(packages, dict):
    k = "gpu" if gpu(mode) else "cpu"
    return packages.get(k, [])

  elif isinstance(packages, list):
    return packages

  else:
    raise argparse.ArgumentTypeError(
        """{}'s "apt_packages" entry must be a dictionary or list, not '{}'""".
        format(CALIBAN_CONFIG, packages))


def caliban_config() -> CalibanConfig:
  """Returns a dict that represents a `.calibanconfig.json` file if present,
  empty dictionary otherwise.

  """
  if not os.path.isfile(CALIBAN_CONFIG):
    return {}

  with open(CALIBAN_CONFIG) as f:
    conf = commentjson.load(f)
    return conf


def expand_experiment_config(items: ExpConf) -> List[Experiment]:
  """Expand out the experiment config for job submission to Cloud.

  """
  if isinstance(items, list):
    return list(
        itertools.chain.from_iterable(
            [expand_experiment_config(m) for m in items]))

  tupleized_items = u.tupleize_dict(items)
  return [u.expand_compound_dict(d) for d in u.dict_product(tupleized_items)]


def validate_compound_keys(m: ExpConf) -> ExpConf:
  """Check that:

  - all key are strings, which do not:
      contain spaces or consecutive commas
      contain commas unless inside a '[...]' compound key
      contain '[' unless at the beginning, and matched by a ']' at the end
      contain ']' unless at the end, and matched by a '[' at the beginning
      begin with '[' and end with ']' unless also containing a comma
  - all values are either boolean, strings, numbers or lists
  """

  def check_k(k):
    if not isinstance(k, str):
      raise argparse.ArgumentTypeError(
          "Key '{}' is invalid! Keys must be strings.".format(k))

    valid_re_str = "[^\s\,\]\[]+"
    list_re = re.compile('\A({}|\[\s*({})(\s*,\s*{})*\s*\])\Z'.format(
        valid_re_str, valid_re_str, valid_re_str))

    if list_re.match(k) is None:
      raise argparse.ArgumentTypeError(
          "Key '{}' is invalid! Not a valid compound key.".format(k))

  def check_v(v):
    types = [list, bool, str, int, float]
    if not any(map(lambda t: isinstance(v, t), types)):
      raise argparse.ArgumentTypeError("Value '{}' in the expanded \
    experiment config '{}' is invalid! Values must be strings, \
    lists, ints, floats or bools.".format(v, m))

  def check_kv_compatibility(k, v):
    """ For already validated k and v, check that
    if k is a compound key, the number of arguments in each sublist must match the
    number of arguments in k """

    if k[0] == '[':
      n_args = len(k.strip('][').split(','))
      if not (isinstance(v, list)):
        raise argparse.ArgumentTypeError(
            "Key '{}' and value '{}' are incompatible: \
                key is compound, but value is not.".format(k, v))
      else:
        if isinstance(v[0], list):
          for vi in v:
            if len(vi) != n_args:
              raise argparse.ArgumentTypeError("Key '{}' and value '{}' have \
                              incompatible arities.".format(k, vi))
        else:
          if len(v) != n_args:
            raise argparse.ArgumentTypeError("Key '{}' and value '{}' have \
                            incompatible arities.".format(k, v))

  if isinstance(m, list):
    return [validate_compound_keys(i) for i in m]

  for k, v in m.items():
    check_k(k)
    check_v(v)
    check_kv_compatibility(k, v)

  return m


def validate_expansion(m: Expansion) -> Expansion:
  """Check that:

  - all key are strings
  - all values are either boolean, strings, numbers or lists
  """

  def valid_k(k):
    return isinstance(k, str)

  def valid_v(v):
    types = [list, bool, str, int, float]
    return any(map(lambda t: isinstance(v, t), types))

  for k, v in m.items():
    if not valid_k(k):
      raise argparse.ArgumentTypeError(
          "Key '{}' is invalid! Keys must be strings.".format(k))

    if not valid_v(v):
      raise argparse.ArgumentTypeError("Value '{}' in the expanded \
experiment config '{}' is invalid! Values must be strings, \
lists, ints, floats or bools.".format(v, m))

  return m


def validate_experiment_config(items: ExpConf) -> ExpConf:
  """Check that the input is either a list of valid experiment configs or a valid
  expansion itself. Returns the list/dict or throws an exception if invalid.

  """

  # Validate the compound keys before expansion
  if isinstance(items, list) or isinstance(items, dict):
    validate_compound_keys(items)
  else:
    raise argparse.ArgumentTypeError("The experiment config is invalid! \
    The JSON file must contain either a dict or a list.")

  for item in expand_experiment_config(items):
    validate_expansion(item)
  return items


def load_experiment_config(s):
  if s.lower() == 'stdin':
    json = commentjson.load(sys.stdin)
  else:
    with open(u.validated_file(s)) as f:
      json = commentjson.load(f)

  return validate_experiment_config(json)


def experiment_to_args(m: Experiment,
                       base: Optional[List[str]] = None) -> List[str]:
  """Returns the list of flag keys and values that corresponds to the supplied
  experiment.

  Keys all expand to the full '--key_name' style that typical Python flags are
  represented by.

  All values except for boolean values are inserted as str(v). For boolean
  values, if the value is True, the key is inserted by itself (in the format
  --key_name). If the value is False, the key isn't inserted at all.

  """
  if base is None:
    base = []

  ret = [] + base

  for k, v in m.items():
    opt = "--{}".format(k)
    if isinstance(v, bool):
      # Append a flag if the boolean flag is true, else do nothing.
      if v:
        ret.append(opt)
    else:
      ret.append("--{}".format(k))
      ret.append(str(v))

  return ret
